qiskit/algorithms/optimizers/cmaes.py

```diff
@@ -0,0 +1,400 @@
+
+    def _initialize_termination_criteria(self) -> None:
+        """
+        Sets the default values for the termination criteria if they haven't been
+        specified by the user. Also removes any entries on the dictionary that don't correspond
+        to a termination criteria.
+        """
+        default_term_crit = {
+            "NoEffectAxis": 0,
+            "TolXUp": 1e4,
+            "TolFun": 1e-12,
+            "TolX": self.sigma_0 * 1e-12,
+            "NoEffectCoord": 0,
+            "ConditionCov": 1e14,
+            "EqualFunValues": 0,
+            "Stagnation": 0,#WIP
+        }
+        # Create default choice of termination criteria if None is passed.
+        if self.termination_criteria is None:
+            self.termination_criteria = default_term_crit
+        # Giving default tolerances to the criteria with None as a parameter.
+        for key, value in self.termination_criteria.items():
+            if key not in default_term_crit:
+                self.termination_criteria.pop(key)
+            if value is None:
+                self.termination_criteria[key] = default_term_crit[key]
+
+    def _initialize_weights(self) -> None:
+        """
+        Initializes the weights of the CMA-ES optimizer.
+        """
+        # Default value for weights.
+        if self.weights is None:
+            self.weights = np.log((self.lmbda + 1) / 2) - np.log(np.arange(1, self.mu + 1))
+
+        # Check if weights are in descending order.
+        if np.any(self.weights[:-1] < self.weights[1:]):
+            raise ValueError("Weights must be in descending order.")
+
+        # Check that all weights are positive.
+        if np.any(self.weights < 0):
+            raise ValueError("Weights must be positive.")
+
+        # Normalize weights.
+        self.weights = self.weights / np.sum(self.weights)
+
+    def _NoEffectAxis(self, tol: float) -> bool:
+        """Termination criteria if the the covariance of a principal axis gets too small."""
+        return 0.1 * np.max(self._state.std_vector) >= np.linalg.norm(self._state.x) * tol
+
+    def _NoEffectCoord(self, tol: float) -> bool:
+        """Termination criteria if the covariance over a cartesian axis gets too small."""
+        return 0.2 * np.max(np.diag(self._state.cov_matrix)) >= np.linalg.norm(self._state.x) * tol
+
+    def _ConditionCov(self, tol: float) -> bool:
+        """What is condition number???"""
+        return True
+
+    def _EqualFunValues(self, tol: float) -> bool:
+        """Termination criteria if the range of the best evaluation during the last 10+ 30*N/lmbda
+        generations is smaller than a given threshold.
+        """
+        if self._state.nit < 10 + 30 * self.N / self.lmbda:
+            return True
+        return max(self._state._best_func_hist) - min(self._state._best_func_hist) > tol
+
+    def _Stagnation(self, tol: float) -> bool:
+        """
+        WIP
```

Comment:
Are you going to sort this out? At present returning True means it always stops - since I am reading these as termination criteria. Since you have Stagnation in the defaults as 0 I an thinking maybe True/False means continue or not since otherwise it would always terminate. From what I can tell these are more whether to continue or not. Since they read as termination criteria I would have more expected them to pass back True when the criteria was met - where the continue logic would handle the logical inverting of the condition - they seem to be more coded like continuation criteria to me